tools/python/m2ctx.py

```diff
@@ -1,71 +1,454 @@
 #!/usr/bin/env python3
 
-import argparse, os, subprocess, sys
+import argparse, os, subprocess, sys, re, time
+from datetime import date
 from pathlib import Path
 
 script_dir = os.path.dirname(os.path.realpath(__file__))
 root_dir = script_dir + "/../../"
 src_dir = root_dir + "src/"
 lib_dir = root_dir + "lib/"
 
+VERSION = "us_1.0"
 
-def get_c_dir(dirname):
-    for root, dirs, files in os.walk(src_dir):
-        for directory in dirs:
-            if directory == dirname:
-                return os.path.join(root, directory)
-    for root, dirs, files in os.walk(lib_dir):
-        for directory in dirs:
-            if directory == dirname:
-                return os.path.join(root, directory)
+ignoreFiles = ["include/sys/regdef.h"]
 
+search_folders = ["include/", "src/"]
 
-def get_c_file(directory):
-    for root, dirs, files in os.walk(directory):
-        for file in files:
-            if file.endswith(".c") and "data" not in file:
-                return file
+# From: https://stackoverflow.com/a/18381470
+# Removes all single line & multi-line comments from a C file.
+def remove_comments(string):
+    pattern = r"(\".*?\"|\'.*?\')|(/\*.*?\*/|//[^\r\n]*$)"
+    # first group captures quoted strings (double or single)
+    # second group captures comments (//single-line or /* multi-line */)
+    regex = re.compile(pattern, re.MULTILINE|re.DOTALL)
+    def _replacer(match):
+        # if the 2nd group (capturing comments) is not None,
+        # it means we have captured a non-quoted (real) comment string.
+        if match.group(2) is not None:
+            return "" # so we will return empty to remove the comment
+        else: # otherwise, we will return the 1st group
+            return match.group(1) # captured quoted-string
+    return regex.sub(_replacer, string)
 
+def regex_get_matches(text, regex):
+    return re.finditer(regex, text, re.MULTILINE)
 
-def import_c_file(in_file):
-    in_file = os.path.relpath(in_file, root_dir)
-    cpp_command = ["gcc", "-E", "-P", "-Iinclude", "-I.", "-Iassets", "-Isrc", "-undef", "-D__sgi", "-D_LANGUAGE_C",
-                   "-DNON_MATCHING", "-D_Static_assert(x, y)=", "-D__attribute__(x)=", in_file]
+# Cleans output to reduce filesize.
+cleanupUselessDefinesRegex = r"#if.*\n#(?:else|define).*\n#endif.*\n"
+def cleanup(out):
+    defs = regex_get_matches(out, cleanupUselessDefinesRegex)
+    for remove in defs:
+        out = out.replace(remove.group(0), '')
+    out = out.replace('\t', ' ') # Replace tabs with spaces.
+    while '\n\n' in out:
+        out = out.replace('\n\n', '\n')
+    while '  ' in out:
+        out = out.replace('  ', ' ')
```

Comment:
Unsure how slow this currently, but one potential future optimisation would be using `re.sub()` to take care of these replacements:
```py
out = re.sub(r"  +", " ", out)
```

Reply:
The program isn't terribly slow. Takes about 1.7 seconds in total, and the majority of that is doing the regex for every single variable in the `get_variable_text()` method. It shouldn't hurt to add this in though.